### PR TITLE
Add precompiles and tests to evm template

### PR DESCRIPTION
**Pull request by fgamundi**
Adds these EVM precompiled smart contracts from Moonkit:

- pallet-balances
- batch
- call-permit
- xcm-utils

Depends on Moonkit's https://github.com/Moonsong-Labs/moonkit/pull/11

### CODE REVIEW

File: container-chains/templates/frontier/runtime/src/migrations.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+        let db_weights = T::DbWeight::get();
+
+        let mut count = 0u64;
+
+        for address in Precompiles::used_addresses() {
+            pallet_evm::Pallet::<T>::create_account(address.into(), revert_bytecode.clone());
+            count += 1;
+        }
+        db_weights.reads_writes(0, count)
```
Comment (fgamundi) [inline container-chains/templates/frontier/runtime/src/migrations.rs:52]: I wasn't sure what weight to return

Comment (tmpolaczyk) [reply container-chains/templates/frontier/runtime/src/migrations.rs:52]: Migration weights are not important until we have multi-block migrations, so this is fine.

Comment (tmpolaczyk) [reply container-chains/templates/frontier/runtime/src/migrations.rs:52]: But looking at the source code of create account I count 2 writes and 1 read, so maybe you can use that?

File: runtime/dancebox/Cargo.toml [hunk 2]
```diff
@@ -234,7 +234,6 @@ try-runtime = [
 	"pallet-invulnerables/try-runtime",
 	"pallet-maintenance-mode/try-runtime",
 	"pallet-migrations/try-runtime",
-	"pallet-migrations/try-runtime",
```
Comment (girazoki) [inline runtime/dancebox/Cargo.toml:237]: reason for this?

Comment (fgamundi) [reply runtime/dancebox/Cargo.toml:237]: It's duplicated, right? The line above is the same

Comment (girazoki) [reply runtime/dancebox/Cargo.toml:237]: ah right

File: Cargo.toml [hunk 3]
```diff
@@ -194,31 +198,32 @@ cumulus-test-relay-sproof-builder = { git = "https://github.com/moondance-labs/p
...
+fc-cli = { git = "https://github.com/moondance-labs/frontier", branch = "tanssi-polkadot-v1.1.0-xcm-util", default-features = false }
+fc-consensus = { git = "https://github.com/moondance-labs/frontier", branch = "tanssi-polkadot-v1.1.0-xcm-util", default-features = false }
+fc-db = { git = "https://github.com/moondance-labs/frontier", branch = "tanssi-polkadot-v1.1.0-xcm-util", default-features = false }
+fc-mapping-sync = { git = "https://github.com/moondance-labs/frontier", branch = "tanssi-polkadot-v1.1.0-xcm-util", default-features = false }
+fc-rpc = { git = "https://github.com/moondance-labs/frontier", branch = "tanssi-polkadot-v1.1.0-xcm-util", features = [
 	"rpc-binary-search-estimate",
 ] }
-fc-rpc-core = { git = "https://github.com/moondance-labs/frontier", branch = "tanssi-polkadot-v1.1.0", default-features = false }
-fc-storage = { git = "https://github.com/moondance-labs/frontier", branch = "tanssi-polkadot-v1.1.0", default-features = false }
+fc-rpc-core = { git = "https://github.com/moondance-labs/frontier", branch = "tanssi-polkadot-v1.1.0-xcm-util", default-features = false }
```
Comment (girazoki) [inline Cargo.toml:225]: the last piece missing, just cherry-pick this in our tanssi-polkadot-v1.1.0 branch

**Review by girazoki — APPROVED** [review]
approved after fixing comments

File: container-chains/templates/frontier/runtime/src/precompiles.rs [hunk 4]
```diff
@@ -14,71 +14,89 @@
...
+        Erc20BalancesPrecompile<R, NativeErc20Metadata>,
+        (CallableByContract, CallableByPrecompile),
+    >,
+    PrecompileAt<
+        AddressU64<2049>,
+        BatchPrecompile<R>,
+        (
+            SubcallWithMaxNesting<2>,
+            // Batch is the only precompile allowed to call Batch.
+            CallableByPrecompile<OnlyFrom<AddressU64<2050>>>,
```
Comment (tmpolaczyk) [inline container-chains/templates/frontier/runtime/src/precompiles.rs:84]:
```suggestion
            CallableByPrecompile<OnlyFrom<AddressU64<2049>>>,
```

Comment (tmpolaczyk) [reply container-chains/templates/frontier/runtime/src/precompiles.rs:84]: Doesn't this have a test?

Comment (girazoki) [reply container-chains/templates/frontier/runtime/src/precompiles.rs:84]: good catch, probably not. @fgamundi can you add a test for this?

Comment (fgamundi) [reply container-chains/templates/frontier/runtime/src/precompiles.rs]: `CallableByPrecompile` rule removed as it isn't needed for batch to call itself. Batch runs the inner calls without changing the original `caller`, so it doesn't count as coming from a precompile. It's the same case for CallPermit.
Added tests for Batch calling itself and CallPermit calling Batch

**Pull request merged**